cpp/lib/src/ZoteroHarvesterUtil.cc

```diff
@@ -288,54 +331,176 @@ bool UploadTracker::urlAlreadyDelivered(const std::string &url, Entry * const en
+                LOG_WARNING("record with URL '" + url + "' already delivered with the same hash (" + record_hash + ")");
+
+            already_delivered = true;
+            break;
+        }
+    }
+
+    if (not already_delivered) {
+        if (hashAlreadyDelivered(record_hash, &hash_bucket, db_connection)) {
+            if (hash_bucket.size() > 1) {
+                LOG_WARNING("multiple records were delivered with the same hash (" + record_hash + ")!");
+                for (const auto &entry : hash_bucket)
+                    LOG_DEBUG(entry.toString());
+            } else
+                LOG_WARNING("record with URL '" + hash_bucket[0].url_ + "' already delivered with the same hash (" + record_hash + ")");
+
+            already_delivered = true;
+        }
     }
 
-    return outdated_journals->size();
+    return already_delivered;
+}
+
+
+bool UploadTracker::urlAlreadyDelivered(const std::string &url, Entry * const entry) const {
+    WaitOnSemaphore lock(&connection_pool_semaphore_);
+    DbConnection db_connection;
+
+    return urlAlreadyDelivered(url, entry, &db_connection);
+}
+
+
+bool UploadTracker::hashAlreadyDelivered(const std::string &hash, std::vector<Entry> * const entries) const {
+    WaitOnSemaphore lock(&connection_pool_semaphore_);
+    DbConnection db_connection;
+
+    return hashAlreadyDelivered(hash, entries, &db_connection);
+}
+
+
+bool UploadTracker::recordAlreadyDelivered(const MARC::Record &record) const {
+    WaitOnSemaphore lock(&connection_pool_semaphore_);
+    DbConnection db_connection;
+
+    const auto hash(Conversion::CalculateMarcRecordHash(record));
+    const auto urls(record.getSubfieldValues("856", 'u'));
+
+    return recordAlreadyDelivered(hash, urls, &db_connection);
 }
 
 
 time_t UploadTracker::getLastUploadTime(const std::string &journal_name) const {
-    db_connection_->queryOrDie("SELECT delivered_at FROM delivered_marc_records WHERE journal_name='" +
-                                db_connection_->escapeString(journal_name) + "' ORDER BY delivered_at DESC");
-    auto result_set(db_connection_->getLastResultSet());
+    WaitOnSemaphore lock(&connection_pool_semaphore_);
+    DbConnection db_connection;
+
+    db_connection.queryOrDie("SELECT delivered_at FROM delivered_marc_records WHERE journal_name='" +
+                             db_connection.escapeString(journal_name) + "' ORDER BY delivered_at DESC");
+    auto result_set(db_connection.getLastResultSet());
     if (result_set.empty())
         return TimeUtil::BAD_TIME_T;
 
     return SqlUtil::DatetimeToTimeT(result_set.getNextRow()["delivered_at"]);
 }
 
 
+bool UploadTracker::archiveRecord(const MARC::Record &record) {
+    WaitOnSemaphore lock(&connection_pool_semaphore_);
+    DbConnection db_connection;
+
+    const auto hash(Conversion::CalculateMarcRecordHash(record));
+    const auto urls(record.getSubfieldValues("856", 'u'));
+    const auto zeder_id(record.getFirstSubfieldValue("ZID", 'a'));
```

Comment:
Generally we want to declare variables as close to their initial use as possible.